apps/api/src/http/internal/rest/webhooks/stripe/invoice.rs

```diff
@@ -0,0 +1,314 @@
+			.metadata
+			.get("USER_ID")
+			.and_then(|i| UserId::from_str(i).ok())
+			.ok_or(TransactionError::custom(ApiError::INTERNAL_SERVER_ERROR))?;
+
+		let status = invoice.status.ok_or(TransactionError::custom(ApiError::BAD_REQUEST))?.into();
+
+		let created_at = invoice
+			.created
+			.and_then(|t| chrono::DateTime::from_timestamp(t, 0))
+			.ok_or(TransactionError::custom(ApiError::BAD_REQUEST))?;
+
+		let db_invoice = Invoice {
+			id: invoice.id.clone().into(),
+			items,
+			customer_id,
+			user_id,
+			paypal_payment_id: None,
+			status,
+			failed: false,
+			refunded: false,
+			disputed: None,
+			created_at,
+			updated_at: created_at,
+			search_updated_at: None,
+		};
+
+		tx.insert_one(db_invoice, None).await?;
+
+		stripe::Invoice::finalize(
+			&global.stripe_client,
+			&invoice.id,
+			FinalizeInvoiceParams {
+				auto_advance: Some(true),
+			},
+		)
+		.await
+		.map_err(|e| {
+			tracing::error!(error = %e, "failed to finalize invoice");
+			TransactionError::custom(ApiError::INTERNAL_SERVER_ERROR)
+		})?;
+	}
+
+	Ok(())
+}
+
+/// Updates the invoice object.
+///
+/// Called for `invoice.updated`, `invoice.finalized`,
+/// `invoice.payment_succeeded`
+pub async fn updated(
+	_global: &Arc<Global>,
+	mut tx: TransactionSession<'_, ApiError>,
+	invoice: stripe::Invoice,
+	_prev_attributes: HashMap<String, serde_json::Value>,
+) -> TransactionResult<(), ApiError> {
+	let id: InvoiceId = invoice.id.into();
+
+	let status: InvoiceStatus = invoice.status.ok_or(TransactionError::custom(ApiError::BAD_REQUEST))?.into();
+
+	let items = invoice_items(invoice.lines.as_ref()).map_err(TransactionError::custom)?;
+
+	tx.update_one(
+		filter::filter! {
+			Invoice {
+				#[query(rename = "_id")]
+				id,
+			}
+		},
+		update::update! {
+			#[query(set)]
+			Invoice {
+				#[query(serde)]
+				status: status,
+				#[query(serde)]
+				items: items,
+				failed: false,
+				updated_at: chrono::Utc::now(),
+			}
+		},
+		None,
+	)
+	.await?;
+
+	Ok(())
+}
+
+/// If invoice is for a subscription, adds a new subscription period to that
+/// subscription. If the subscription is still in trial, makes the period a
+/// trial period. Updates the invoice object.
+///
+/// Called for `invoice.paid`
+pub async fn paid(
+	global: &Arc<Global>,
+	mut tx: TransactionSession<'_, ApiError>,
+	invoice: stripe::Invoice,
+) -> TransactionResult<Option<SubscriptionId>, ApiError> {
+	if let Some(subscription) = &invoice.subscription {
+		let items = invoice_items(invoice.lines.as_ref())
+			.map_err(TransactionError::custom)?
+			.into_iter()
+			.collect::<Vec<_>>();
+
+		let products = tx
+			.find(
+				filter::filter! {
+					SubscriptionProduct {
+						#[query(flatten)]
+						variants: SubscriptionProductVariant {
+							#[query(selector = "in", serde)]
+							id: items,
+						}
+					}
+				},
+				None,
+			)
+			.await?;
+
+		if products.len() != 1 {
+			// only accept invoices for one of our products
+			return Ok(None);
```

Comment:
This path avoids the invoice update at the end, perhaps we should do that before we check the subscription?

Reply:
Did you cover this?